File: 1536. minimum-swaps-to-arrange-a-binary-grid.py
class Solution:
    def minSwaps(self, grid):
        n = len(grid)
        zeros = []

        for row in grid:
            count = 0
            for j in range(n - 1, -1, -1):
                if row[j] == 0:
                    count += 1
                else:
                    break
            zeros.append(count)

        swaps = 0

        for i in range(n):
            needed = n - i - 1
            j = i
            while j < n and zeros[j] < needed:
                j += 1
            if j == n:
                return -1
            while j > i:
                zeros[j], zeros[j - 1] = zeros[j - 1], zeros[j]
                j -= 1
                swaps += 1

        return swaps

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1:
    grid = [[0,0,1],[1,1,0],[1,0,0]]
    output = sol.minSwaps(grid)
    print(f"Example 1: Input: grid = {grid}")
    print(f"Output: {output}")
    print(f"Expected: 3")
    print()
    
    # Example 2:
    grid = [[0,1,1,0],[0,1,1,0],[0,1,1,0],[0,1,1,0]]
    output = sol.minSwaps(grid)
    print(f"Example 2: Input: grid = {grid}")
    print(f"Output: {output}")
    print(f"Expected: -1")
    print()
    
    # Example 3:
    grid = [[1,0,0],[1,1,0],[1,1,1]]
    output = sol.minSwaps(grid)
    print(f"Example 3: Input: grid = {grid}")
    print(f"Output: {output}")
    print(f"Expected: 0")

File: 1536. minimum-swaps-to-arrange-a-binary-grid.java
class Solution {
    public int minSwaps(int[][] grid) {
        int n = grid.length;
        int[] zeros = new int[n];
        for (int i = 0; i < n; i++) {
            int count = 0;
            for (int j = n - 1; j >= 0 && grid[i][j] == 0; j--) {
                count++;
            }
            zeros[i] = count;
        }

        int swaps = 0;

        for (int i = 0; i < n; i++) {
            int needed = n - i - 1;
            int j = i;
            while (j < n && zeros[j] < needed) j++;

            if (j == n) return -1;
            while (j > i) {
                int temp = zeros[j];
                zeros[j] = zeros[j - 1];
                zeros[j - 1] = temp;
                j--;
                swaps++;
            }
        }

        return swaps;
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1:
        int[][] grid1 = {{0,0,1},{1,1,0},{1,0,0}};
        int output1 = sol.minSwaps(grid1);
        System.out.println("Example 1: Input: grid = [[0,0,1],[1,1,0],[1,0,0]]");
        System.out.println("Output: " + output1);
        System.out.println("Expected: 3");
        System.out.println();
        
        // Example 2:
        int[][] grid2 = {{0,1,1,0},{0,1,1,0},{0,1,1,0},{0,1,1,0}};
        int output2 = sol.minSwaps(grid2);
        System.out.println("Example 2: Input: grid = [[0,1,1,0],[0,1,1,0],[0,1,1,0],[0,1,1,0]]");
        System.out.println("Output: " + output2);
        System.out.println("Expected: -1");
        System.out.println();
        
        // Example 3:
        int[][] grid3 = {{1,0,0},{1,1,0},{1,1,1}};
        int output3 = sol.minSwaps(grid3);
        System.out.println("Example 3: Input: grid = [[1,0,0],[1,1,0],[1,1,1]]");
        System.out.println("Output: " + output3);
        System.out.println("Expected: 0");
    }
}